pkg/apis/core/validation/validation.go

```diff
@@ -3144,6 +3158,15 @@ func validateInitContainers(containers []core.Container, regularContainers []cor
 		// Apply the validation common to all container types
 		allErrs = append(allErrs, validateContainerCommon(&ctr, volumes, podClaimNames, idxPath, opts)...)
 
+		// Apply the validation specific to init containers
+		if ctr.RestartPolicy != nil {
+			if opts.AllowSidecarContainers {
+				allErrs = append(allErrs, validateInitContainerRestartPolicy(ctr.RestartPolicy, idxPath.Child("restartPolicy"))...)
+			} else {
+				allErrs = append(allErrs, field.Forbidden(idxPath.Child("restartPolicy"), "may not be set for init containers without the SidecarContainers feature"))
+			}
+		}
```

Comment:
When `opts.AllowSidecarContainers` is enabled, a pod is created with a `RestartPolicy: Always` init container, and `opts.AllowSidecarContainers` is disabled again, the pod needs to remain usable.

Instead of checking `opts.AllowSidecarContainers` let's follow the steps defined in [API changes: New field in existing API version](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) and drop writing new values to the field before writing to storage if the feature gate is off

Reply:
> is disabled again, the pod needs to remain usable.

It cannot remain usable. Ignoring the field will make Pod never start. This is what we wrote in KEP PRR: https://github.com/kubernetes/enhancements/tree/512b6c8ed83e2107e04b1588621a006c7b5c3cf0/keps/sig-node/753-sidecar-containers#feature-enablement-and-rollback

Reply:
perhaps we need a test for this.

Reply:
Ah, you're right. Should updates to this object be allowed if the value of this field is unchanged?  I.e. for update, should oldRestartPolicy=currentRestartPolicy updates be allowed? Or is it fine to just make all updates to the Pod be forced to fail.

Reply:
I'd say force to fail. At least in Alpha

Reply:
Ok. Just the comment here for future API reviewers and we should be good.

Reply:
commented.

Reply:
> Instead of checking `opts.AllowSidecarContainers` let's follow the steps defined in [API changes: New field in existing API version](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#new-field-in-existing-api-version) and drop writing new values to the field before writing to storage if the feature gate is off

applied it for consistency with other fields.